src/main/java/dev/blaauwendraad/masker/json/KeyContainsMasker.java

```diff
@@ -174,28 +164,101 @@ public byte[] mask(byte[] input) {
+     * In case the current JsonPATH segment is an array segment, the method checks if the current element needs to be masked.
      */
     private void trackCurrentJsonPath(MaskingState maskingState) {
-        // Check if this is the end of the current json object
-        if (!maskingState.isInArray() && currentByteIsUnescapedCurlyBracketClose(maskingState)) {
-            maskingState.backtrackCurrentJsonPath();
+        // special case of empty json objects: look ahead of the curly brackets opening to check if the key is an empty
+        // json object. If so, put a placeholder onto the current json path stack. If no, rollback to the current maskingState index.
+        if (isCurlyBracketOpen(maskingState.byteAtCurrentIndex()) && !maskingState.isInArraySegment()) {
+            int currentIndex = maskingState.currentIndex();
+            maskingState.incrementCurrentIndex();
+            skipWhitespaceCharacters(maskingState);
+            if (isCurlyBracketClose(maskingState.byteAtCurrentIndex())) {
+                maskingState.expandCurrentJsonPath(maskingState.byteAtCurrentIndex(), 1);
+            } else {
+                maskingState.setCurrentIndex(currentIndex);
+            }
         }
-        // Check if this is the end of the current json array
-        if (maskingState.isInArray() && currentByteIsUnescapedSquareBracketClose(maskingState)) {
+        // Check if this is the end of the current non-empty json object.
+        if (currentByteIsUnescapedCurlyBracketClose(maskingState) && !maskingState.isInRootSegment() && !maskingState.isInArraySegment()) {
             maskingState.backtrackCurrentJsonPath();
-            // if it is not a nested array, backtrack the parent key component as well
-            if (!maskingState.isInArray()) {
-                maskingState.backtrackCurrentJsonPath();
-            }
         }
-        // Check if this is the end of the current json array element
-        if (maskingState.isInArray() && currentByteIsUnescapedComma(maskingState)) {
-            maskingState.incrementCurrentJsonPathArrayIndex();
+        // Check if this is the end of the current json array.
+        if (currentByteIsUnescapedSquareBracketClose(maskingState) && maskingState.isInArraySegment()) {
+            maskingState.backtrackCurrentJsonPath();
         }
         // Check if this is the start of a json array
         if (currentByteIsUnescapedSquareBracketOpen(maskingState)) {
             maskingState.expandCurrentJsonPath();
+            // Check is the first element in the array has to be masked.
+            KeyMaskingConfig keyMaskingConfig = keyMatcher.getMaskConfigIfMatched(maskingState.getMessage(), maskingState.getCurrentJsonPath());
+            if (keyMaskingConfig != null) {
+                maskArrayElementInPlace(maskingState, keyMaskingConfig);
+                trackCurrentJsonPath(maskingState);
```

Comment:
Does it make sense to move it inside `maskArrayElementInPlace`?

Reply:
While looking more into it and this way of tracking makes it quite confusing - we have explicit expanding and backtracking in some places, but implicit inside `trackCurrentJsonPath` that does it based on the current state. Isn't this particular occurrence is simply `maskingState.incrementCurrentJsonPathArrayIndex();`?

Reply:
> Isn't this particular occurrence is simply maskingState.backtrackCurrentJsonPath();

No really. It could be expanding to a new array segment (next array element is an array), incrementing the index in the current array segment, backtracking current array segment or exapnding to a new object segment (next array element is an object).

> Does it make sense to move it inside maskArrayElementInPlace?

I kept it in the track method so that "maskArrayElementInPlace" is independent of the tracking logic.
In general, I moved as much tracking logic as possible inside "trackCurrentJsonPath" method. This way it is easier to disable it if there are no jsonpath keys.

Reply:
Okay, I thought it was more straightforward